### PR TITLE
chore: typescript

### PR DESCRIPTION
**Pull request by holloway**
Converts JS to TS

There are a few remaining TS errors, and parts to [continue refactoring per the Vue TS docs](https://vuejs.org/guide/typescript/composition-api.html), however these don't affect the build, and this PR is getting quite large so it's probably better to review as-is, and continue in another PR.

### CODE REVIEW

File: client/app.vue [hunk 1]
```diff
@@ -17,7 +17,9 @@
   </div>
 </template>
 
-<script setup>
+<script setup lang="ts">
+import { overlayModalKey, overlayModalMethodsKey } from './providers/providerKeys';
```
Comment (NGPixel) [inline client/app.vue:21]: No semicolons!

File: client/components/AssignmentTrayItem.vue [hunk 2]
```diff
@@ -18,14 +18,23 @@
   </div>
 </template>
 
-<script setup>
+<script setup lang="ts">
 import { inject } from 'vue'
+import { deleteAssignmentKey } from '~/providers/providerKeys';
```
Comment (NGPixel) [inline client/components/AssignmentTrayItem.vue:23]: No semicolons!

File: client/components/AssignmentTrayItem.vue [hunk 3]
```diff
@@ -18,14 +18,23 @@
...
+    person?: {
+      name: string
+    }
+    role: string
+  }
+}>()
 
-const deleteAssignment = inject('deleteAssignment')
+const _deleteAssignment = inject(deleteAssignmentKey)
+if(!_deleteAssignment) throw Error("Sdfsdf")
```
Comment (NGPixel) [inline client/components/AssignmentTrayItem.vue:36]: Avoid single line if, use `{ }`. It's more verbose but it's clearer.

File: client/components/AssignmentTrayItem.vue [hunk 4]
```diff
@@ -18,14 +18,23 @@
...
+const deleteAssignment= _deleteAssignment
```
Comment (NGPixel) [inline client/components/AssignmentTrayItem.vue:37]: Missing space before `=`

File: client/components/Button.vue [hunk 5]
```diff
@@ -5,13 +5,12 @@
...
-  btnType: {
-    type: String,
-    default: 'default'
-  }
-})
+<script setup lang="ts">
+export type Props = { btnType: keyof (typeof classForBtnType) }
+const props = withDefaults(
+  defineProps<Props>(),
+  { btnType: 'default' },
```
Comment (NGPixel) [inline client/components/Button.vue:12]: No dangling comma

File: client/components/DocumentCard.vue [hunk 6]
```diff
@@ -55,20 +55,27 @@ Based on https://tailwindui.com/components/application-ui/lists/grid-lists#compo
   </li>
 </template>
 
-<script setup>
+<script setup lang="ts">
 import { inject } from 'vue'
 import { DateTime } from 'luxon'
+import { assignEditorKey } from '~/providers/providerKeys';
+import type { ADocument } from './DocumentCardsTypes';
```
Comment (NGPixel) [inline client/components/DocumentCard.vue:62]: No semicolons!

File: client/components/DocumentCard.vue [hunk 7]
```diff
@@ -55,20 +55,27 @@ Based on https://tailwindui.com/components/application-ui/lists/grid-lists#compo
...
+type Props = {
+  document: ADocument
+  selected?: boolean
+}
 
-const assignEditor = inject('assignEditor')
+const props = defineProps<Props>()
+
+const _assignEditor = inject(assignEditorKey)
+if(!_assignEditor) throw Error("Required assignEditor provider")
```
Comment (NGPixel) [inline client/components/DocumentCard.vue:72]: No single-line if

File: client/components/DocumentCards.vue [hunk 8]
```diff
@@ -9,23 +9,27 @@ Based on https://tailwindui.com/components/application-ui/lists/grid-lists#compo
   </ul>
 </template>
 
-<script setup>
+<script setup lang="ts">
 import { provide } from 'vue'
+import { assignEditorKey, deleteAssignmentKey } from '~/providers/providerKeys';
+import type { ADocument } from './DocumentCardsTypes';
```
Comment (NGPixel) [inline client/components/DocumentCards.vue:15]: No semicolons!

File: client/components/HeaderNav.vue [hunk 9]
```diff
@@ -161,8 +161,11 @@ const { data: allUsers } = await useAsyncData(
   {
     default: () => ([]),
     server: false,
-    transform: (resp) => resp.toSorted((a, b) => a.name.localeCompare(b.name))
+    transform: (resp) => resp?.toSorted((a, b) => a.name.localeCompare(b.name))
   }
 )
 
+if (!_allUsers || !Array.isArray(_allUsers)) throw Error("Required allUsers data")
```
Comment (NGPixel) [inline client/components/HeaderNav.vue:168]: Avoid single-line if

File: client/components/RpcLabelEditDialog.vue [hunk 10]
```diff
@@ -68,19 +68,24 @@
...
+<script setup lang="ts">
 import { ColorEnum } from '~/rpctracker_client'
+import type { Label } from '~/rpctracker_client'
+import { overlayModalMethodsKey } from '../providers/providerKeys'
 
 const api = useApi()
 
-const { ok, cancel } = inject('overlayModalMethods')
+const _overlayModalMethods = inject(overlayModalMethodsKey)
+if(!_overlayModalMethods) throw Error("Expected provider of overlayModalMethods")
```
Comment (NGPixel) [inline client/components/RpcLabelEditDialog.vue:79]: Avoid single-line if

File: client/components/RpcLabelPicker.vue [hunk 11]
```diff
@@ -34,31 +36,28 @@
   </HeadlessCombobox>
 </template>
 
-<script setup>
-const props = defineProps({
+<script setup lang="ts">
+import type { Label } from '~/rpctracker_client';
```
Comment (NGPixel) [inline client/components/RpcLabelPicker.vue:40]: No semicolons!

File: client/components/TestOverlay.vue [hunk 12]
```diff
@@ -31,14 +31,17 @@
   </div>
 </template>
 
-<script setup>
-const { ok, cancel } = inject('overlayModalMethods')
+<script setup lang="ts">
+import { overlayModalMethodsKey } from '~/providers/providerKeys';
+
+const overlayModalMethods = inject(overlayModalMethodsKey) 
+if(!overlayModalMethods) throw Error('overlayModalMethods used outside provider');
```
Comment (NGPixel) [inline client/components/TestOverlay.vue:38]: No semicolons!

File: client/components/UserCreateDialog.vue [hunk 13]
```diff
@@ -127,15 +127,29 @@
...
 
-<script setup>
+<script setup lang="ts">
+import { overlayModalMethodsKey } from '~/providers/providerKeys';
 
 // DIALOG
 
-const { ok, cancel } = inject('overlayModalMethods')
+const overlayModalMethods = inject(overlayModalMethodsKey) 
+if(!overlayModalMethods) throw Error('overlayModalMethods used outside provider');
```
Comment (NGPixel) [inline client/components/UserCreateDialog.vue:136]: No semicolons!

File: client/pages/auth48/[section].vue [hunk 14]
```diff
@@ -26,7 +26,9 @@
   <p class="pt-5 text-center text-gray-500">(data are mocked)</p>
 </template>
 
-<script setup>
+<script setup lang="ts">
+import type { Column } from '~/components/DocumentTableTypes';
```
Comment (NGPixel) [inline client/pages/auth48/[section].vue:30]: No semicolons!

File: client/pages/docs/[id].vue [hunk 15]
```diff
@@ -268,10 +268,10 @@ const { data: people } = await useAsyncData(
   { server: false, default: () => [] }
 )
 
-async function saveLabels (labels) {
+async function saveLabels (labels: number[]) {
   if (!draftPending.value) {
     await api.documentsPartialUpdate({
-      draftName: draft.value.name,
+      draftName: draft.value?.name ?? "",
```
Comment (NGPixel) [inline client/pages/docs/[id].vue:274]: Replace with single quotes

File: client/pages/docs/assignments.vue [hunk 16]
```diff
@@ -58,8 +58,9 @@
   </div>
 </template>
 
-<script setup>
+<script setup lang="ts">
 import { DateTime } from 'luxon'
+import type { Assignment } from '~/rpctracker_client';
```
Comment (NGPixel) [inline client/pages/docs/assignments.vue:63]: No semicolons!

File: client/pages/docs/assignments.vue [hunk 17]
```diff
@@ -120,15 +121,15 @@ const currentFilterDesc = computed(() => {
...
-      rfc_to_be: assignment.rfcToBeId,
-      person: assignment.personId,
-      role: documents.value.find(d => d.id === assignment.rfcToBeId)?.needsAssignment?.slug ?? 'first_editor'
+      rfc_to_be: assignment.rfcToBe,
+      person: assignment.person,
+      role: documents.value.find((d: any) => d.id === assignment.rfcToBe)?.needsAssignment?.slug ?? 'first_editor'
     },
     method: 'POST',
-    headers: { 'X-CSRFToken': csrf.value }
+    headers: { 'X-CSRFToken': csrf.value?.toString() ?? "" }
```
Comment (NGPixel) [inline client/pages/docs/assignments.vue:132]: Replace with single quotes

File: client/pages/docs/assignments.vue [hunk 18]
```diff
@@ -143,10 +144,10 @@ function compareEditors (a, b) {
   return comparisons.find(c => c !== 0) ?? 0
 }
 
-async function deleteAssignment (assignment) {
+async function deleteAssignment (assignment: Assignment) {
   await $fetch(`/api/rpc/assignments/${assignment.id}`, {
     method: 'DELETE',
-    headers: { 'X-CSRFToken': csrf.value }
+    headers: { 'X-CSRFToken': csrf?.value ?? "" }
```
Comment (NGPixel) [inline client/pages/docs/assignments.vue:150]: Replace with single quotes

File: client/pages/index.vue [hunk 19]
```diff
@@ -35,8 +35,9 @@
   </div>
 </template>
 
-<script setup>
+<script setup lang="ts">
 import { TestOverlay } from '#components'
+import { overlayModalKey } from '~/providers/providerKeys';
```
Comment (NGPixel) [inline client/pages/index.vue:40]: No semicolons!

File: client/pages/index.vue [hunk 20]
```diff
@@ -59,7 +60,9 @@ const stats = [
 
 // OVERLAY TEST
 
-const { openOverlayModal } = inject('overlayModal')
+const _overlayModal = inject(overlayModalKey)
+if(!_overlayModal) throw Error("Require provider overlayModalKey")
```
Comment (NGPixel) [inline client/pages/index.vue:64]: Avoid single-line if

File: client/pages/queue/[section].vue [hunk 21]
```diff
@@ -43,11 +43,12 @@
...
 </template>
 
-<script setup>
+<script setup lang="ts">
 import { DateTime } from 'luxon'
 import Fuse from 'fuse.js/basic'
 import { useSiteStore } from '@/stores/site'
-import Badge from '../components/Badge'
+import Badge from '../../components/Badge.vue'
+import type { Column } from '~/components/DocumentTableTypes';
```
Comment (NGPixel) [inline client/pages/queue/[section].vue:51]: No semicolons!

File: client/pages/stats.vue [hunk 22]
```diff
@@ -41,6 +45,11 @@ const { data: labels } = await useAsyncData(() => api.labelsList(), { server: fa
 
 const { data: documents } = await useAsyncData(() => api.documentsList(), { server: false, default: () => [] })
 
-const { data: labelStats } = await useAsyncData(() => api.statsLabels(), { server: false, default: () => [] })
+const { data: labelStats } = await useAsyncData(
+  () => api.statsLabels(), {
+    server: false,
+    default: () => ({ labelStats: [] }) as LabelStats,
```
Comment (NGPixel) [inline client/pages/stats.vue:51]: No dangling comma

File: client/pages/team/index.vue [hunk 23]
```diff
@@ -40,11 +40,14 @@
   </div>
 </template>
 
-<script setup>
+<script setup lang="ts">
 import { UserCreateDialog } from '#components'
+import { overlayModalKey } from '~/providers/providerKeys';
```
Comment (NGPixel) [inline client/pages/team/index.vue:45]: No semicolons!

File: client/pages/team/index.vue [hunk 24]
```diff
@@ -40,11 +40,14 @@
...
 
 const snackbar = useSnackbar()
-const { openOverlayModal } = inject('overlayModal')
+const _overlayModal = inject(overlayModalKey)
+if(!_overlayModal) throw Error("_overlayModal must be used in provider")
```
Comment (NGPixel) [inline client/pages/team/index.vue:49]: Avoid single-line if

File: client/providers/providerKeys.ts [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+        component: null | VNode | Component,
+        componentProps?: null | Record<string, unknown>,
+        mode?: "overlay" | "side"
+    }) => Promise<void>
+    closeOverlayModal: () => void
+}>
+
+export const overlayModalMethodsKey = Symbol() as InjectionKey<{
+    ok: (val?: string) => void,
+    cancel: () => void,
```
Comment (NGPixel) [inline client/providers/providerKeys.ts:17]: No dangling comma

**Review by NGPixel — CHANGES_REQUESTED** [review]

File: client/.eslintrc.js [hunk 26]
```diff
@@ -1,3 +1,6 @@
+import eslint from '@eslint/js';
+import tseslint from 'typescript-eslint';
```
Comment (NGPixel) [inline client/.eslintrc.js:2]: No semicolons!

File: client/.eslintrc.js [hunk 27]
```diff
@@ -9,9 +9,11 @@ module.exports = {
   },
   extends: [
     'standard',
-    'plugin:nuxt/recommended'
+    'plugin:nuxt/recommended',
     // 'plugin:vue/vue3-essential', // Priority A: Essential (Error Prevention)
     // 'plugin:vue/vue3-strongly-recommended' // Priority B: Strongly Recommended (Improving Readability)
+    'plugin:@typescript-eslint/recommended',
+    'plugin:@typescript-eslint/stylistic',
```
Comment (NGPixel) [inline client/.eslintrc.js:16]: No dangling comma

File: client/components/RpcLabel.vue [hunk 28]
```diff
@@ -5,14 +5,13 @@
...
-  /**
-   * Label to be rendered
-   */
-  label: {
-    type: Object,
-    required: true
-  }
-})
+<script setup lang="ts">
+import type { Label } from '~/rpctracker_client';
```
Comment (NGPixel) [inline client/components/RpcLabel.vue:9]: No semicolons!

**Review by NGPixel — CHANGES_REQUESTED** [review]
More linting police

File: client/.eslintrc.js [hunk 29]
```diff
@@ -9,14 +9,18 @@ module.exports = {
...
     // 'plugin:vue/vue3-strongly-recommended' // Priority B: Strongly Recommended (Improving Readability)
+    'plugin:@typescript-eslint/recommended',
+    'plugin:@typescript-eslint/stylistic'
   ],
   globals: {
     d3: true
   },
+  parser: 'vue-eslint-parser',
   parserOptions: {
+    parser: '@typescript-eslint/parser',
```
Comment (holloway) [inline client/.eslintrc.js:23]: This seems to fix the broken linting fyi @NGPixel

**Comment by holloway** [comment]
@NGPixel ready for a (hopefully final!) review

**Comment by rjsparks** [comment]
All - please collapse resolved comments

**Review by NGPixel — APPROVED** [review]

**Comment by holloway** [comment]
@jennifer-richards any chance this can be merged so that I can continue on Monday? Thanks!

**Pull request merged**